### PR TITLE
Move to Crowdin Api v2 and dotnet 8

### PR DESCRIPTION
**Pull request by jasonleenaylor**
* Refactor and reuse three classes from Paratext to interact with v2 (CrowdinHelperBase, DownloadCrowdInHelper, UploadCrowdinHelper)
* Drop some commands and options that are no longer needed or that were never used

+semver:major

- - -
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/overcrowdin/49)

### CODE REVIEW

File: src/Overcrowdin/CrowdInHelperBase.cs [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+			{
+				Console.Error.WriteLine("Failed to get download URL");
+				return false;
+			}
+
+			// Download the build translation and save it to disk
+			var httpClient = _httpClientFactory.GetClient();
+			using (var zipFileStream = _fileSystem.FileStream.New(outputPath, FileMode.Create)) // Can't use a "using" statement since it gets disposed before the enumeration is read
+			using (var stream = await httpClient.GetStreamAsync(url))
+				stream.CopyTo(zipFileStream);
```
Comment (hahn-kev) [inline src/Overcrowdin/CrowdInHelperBase.cs:188]: I'd use the async version here

File: src/Overcrowdin/CrowdInHelperBase.cs [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+			Console.WriteLine("Translation build finished");
+			return build.Id;
+		}
+
+		private async Task<ProjectBuild> WaitForBuildToComplete(ProjectBuild build)
+		{
+			int prevProgressDiv10 = 0;
+			do
+			{
+				Thread.Sleep(2000);
```
Comment (hahn-kev) [inline src/Overcrowdin/CrowdInHelperBase.cs:304]: use Task.Delay instead of Thread sleep

File: src/Overcrowdin/CrowdInHelperBase.cs [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+			{
+				Console.Error.WriteLine("Translation build stopped unexpectedly: " + build.Status);
+				return null;
+			}
+
+			Console.WriteLine("Translation build finished");
+			return build.Id;
+		}
+
+		private async Task<ProjectBuild> WaitForBuildToComplete(ProjectBuild build)
```
Comment (hahn-kev) [inline src/Overcrowdin/CrowdInHelperBase.cs:299]: there should probably be some kind of timeout for this check, at the moment it looks like it'll just go forever.

File: src/Overcrowdin/CrowdinProjectSettings.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+				{
+					var sourceExecutor = new SourceFilesApiExecutor(apiInstance);
+					var branches = await sourceExecutor.ListBranches(settings.ProjectId);
+					var branch = branches.Data.FirstOrDefault(b => b.Name == branchName);
+					settings.BranchId = branch?.Id ?? 0;
+				}
+			}
+			catch
+			{
+				throw new Exception($"Could not find project with name {project} and branch {branchName}");
```
Comment (hahn-kev) [inline src/Overcrowdin/CrowdinProjectSettings.cs:42]: I'd pass the inner exception here, so you can track down what's going on, maybe there's a different error than you assume

File: src/Overcrowdin/UpdateCommand.cs [hunk 5]
```diff
@@ -23,18 +23,15 @@ public class Options : GlobalOptions, IFileOptions
...
 			{
 				return 1;
 			}
-
-			var updateFileParametersList = new List<UpdateFileParameters>();
+			var uploadHelper = await CrowdInUploadHelper.Create(credentials, fileSystem, apiFactory);
+			var updateFileParametersList = new List<FileParameters>();
 			CommandUtilities.GetFileList(config, opts, fileSystem, updateFileParametersList, new SortedSet<string>());
 
 			if (!updateFileParametersList.Any())
```
Comment (hahn-kev) [inline src/Overcrowdin/UpdateCommand.cs:37]: I would move this near the top of this method. No need to fetch credentials and make an upload helper if there's no files to add.

File: src/Overcrowdin/CommandUtilities.cs [hunk 6]
```diff
@@ -30,29 +29,10 @@ public static Credentials GetCredentialsFromConfiguration(IConfiguration config)
...
-			if (string.IsNullOrEmpty(userName))
-			{
-				return new ProjectCredentials {ProjectKey = apiKey};
-			}
-			return new AccountCredentials { AccountKey = apiKey, LoginName = userName };
+			var settings = await CrowdinProjectSettings.Init(config["project_identifier"], branch, apiKey, apiFactory);
+			return settings;
 		}
 
 		public static void GetFileList<T>(IConfiguration config, IFileOptions opts, IFileSystem fs,
```
Comment (hahn-kev) [inline src/Overcrowdin/CommandUtilities.cs:38]: I don't see the folders parameter being used anywhere this method is called, it could probably be removed.

**Review by hahn-kev — CHANGES_REQUESTED** [review]
I noticed most of your test methods where `async void` when they should be `async Task`, looks like I don't have push permission otherwise I would have fixed it.

I left a few other comments of misc stuff I noticed that could be better. I'll be honest I didn't read every line of code though, it's a lot.

**Review by jasonleenaylor — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/sillsdev/overcrowdin/49)* status: 0 of 45 files reviewed, 6 unresolved discussions (waiting on @hahn-kev)

___
*[`src/Overcrowdin/CommandUtilities.cs` line 38 at r1](https://reviewable.io/reviews/sillsdev/overcrowdin/49#gh-1563304344:-NvJWpeAEFvbrR3QYF07:b-e4xml4) ([raw file](https://github.com/sillsdev/overcrowdin/blob/2fac8f9d5e3b6f0e86a0b051678ea4c419b32ec3/src/Overcrowdin/CommandUtilities.cs#L38)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

I don't see the folders parameter being used anywhere this method is called, it could probably be removed.
</blockquote></details>

Line 54.
___
*[`src/Overcrowdin/CrowdInHelperBase.cs` line 188 at r1](https://reviewable.io/reviews/sillsdev/overcrowdin/49#gh-1563284528:-NvJW-lTAvFB46Ogl1Hy:b-896fix) ([raw file](https://github.com/sillsdev/overcrowdin/blob/2fac8f9d5e3b6f0e86a0b051678ea4c419b32ec3/src/Overcrowdin/CrowdInHelperBase.cs#L188)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

I'd use the async version here
</blockquote></details>

Done.
___
*[`src/Overcrowdin/CrowdInHelperBase.cs` line 299 at r1](https://reviewable.io/reviews/sillsdev/overcrowdin/49#gh-1563292819:-NvJXF_p2ow-L49jOSUV:bb6un1j) ([raw file](https://github.com/sillsdev/overcrowdin/blob/2fac8f9d5e3b6f0e86a0b051678ea4c419b32ec3/src/Overcrowdin/CrowdInHelperBase.cs#L299)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

there should probably be some kind of timeout for this check, at the moment it looks like it'll just go forever.
</blockquote></details>

Done
___
*[`src/Overcrowdin/CrowdInHelperBase.cs` line 304 at r1](https://reviewable.io/reviews/sillsdev/overcrowdin/49#gh-1563291556:-NvJX6zDDCxiNzFp3mgm:b-896fix) ([raw file](https://github.com/sillsdev/overcrowdin/blob/2fac8f9d5e3b6f0e86a0b051678ea4c419b32ec3/src/Overcrowdin/CrowdInHelperBase.cs#L304)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

use Task.Delay instead of Thread sleep
</blockquote></details>

Done.
___
*[`src/Overcrowdin/CrowdinProjectSettings.cs` line 42 at r1](https://reviewable.io/reviews/sillsdev/overcrowdin/49#gh-1563294197:-NvJZYHBAnWFzHJg7rRk:b-896fix) ([raw file](https://github.com/sillsdev/overcrowdin/blob/2fac8f9d5e3b6f0e86a0b051678ea4c419b32ec3/src/Overcrowdin/CrowdinProjectSettings.cs#L42)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

I'd pass the inner exception here, so you can track down what's going on, maybe there's a different error than you assume
</blockquote></details>

Done.
___
*[`src/Overcrowdin/UpdateCommand.cs` line 37 at r1](https://reviewable.io/reviews/sillsdev/overcrowdin/49#gh-1563299869:-NvJa2MJ3xQJ5pT_3bFq:b-896fix) ([raw file](https://github.com/sillsdev/overcrowdin/blob/2fac8f9d5e3b6f0e86a0b051678ea4c419b32ec3/src/Overcrowdin/UpdateCommand.cs#L37)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

I would move this near the top of this method. No need to fetch credentials and make an upload helper if there's no files to add.
</blockquote></details>

Done.

**Pull request merged**